models/credmark/protocols/lending/compound/compound_v2.py

```diff
@@ -59,51 +103,103 @@ def run(self, input: Token) -> dict:
         totalSupply = comp_asset[3]/pow(10, decimals)
         totalBorrows = comp_asset[6]/pow(10, decimals)
         output[symbol] = {'totalsupply': totalSupply, 'totalborrow': totalBorrows}
 
         return output
 
 
-@Model.describe(slug="compound-token-liability",
+class CompoundDebtInfo(DTO):
+    inputSymbol: str
+    tokenName: str
+    cTokenName: str
+    token: Token
+    cToken: Token
+    totalLiquidity: float
+    totalBorrows: float
+    totalReserves: float
+
+
+class CompoundDebtInfos(IterableListGenericDTO[CompoundDebtInfo]):
+    compoundDebtInfos: List[CompoundDebtInfo]
+    _iterator: str = 'CompoundDebtInfos'
+
+
+@Model.describe(slug="compound.all-liability",
                 version="1.0",
                 display_name="Compound V2 token liability",
                 description="Compound V2 token liability at a given block number",
-                input=Token)
-class CompoundV2GetTokenLiability(Model):
-    def run(self, input: Token) -> dict:
-        if not input.address:
-            raise ModelRunError(f'Input token is invalid, {input}')
+                input=EmptyInput,
+                output=CompoundDebtInfos)
+class CompoundV2TotalLiability(Model):
 
-        output = {}
-        tokenContract = Contract(
-            address=input.address.checksum,
-            abi=ERC_20_TOKEN_CONTRACT_ABI)
+    def run(self, input: EmptyInput) -> CompoundDebtInfos:
+
+        debts = []
+
+        for _, tokenAddress in COMPOUND_CTOKEN.items():
+            debt = self.context.run_model(slug='compound.token-liability', input=Token(address=tokenAddress))
+            debts.append(debt)
+
+        return CompoundDebtInfos(compoundDebtInfos=debts)
+
+
+@Model.describe(slug="compound.token-liability",
+                version="1.0",
+                display_name="Compound V2 token liability",
+                description="Compound V2 token liability at a given block number",
+                input=Token,
+                output=CompoundDebtInfo)
+class CompoundV2GetTokenLiability(Model):
+    def run(self, input: Token) -> CompoundDebtInfo:
+        cToken = Token(address=input.address,
+                       abi=COMPOUND_CTOKEN_CONTRACT_ABI)
+
+        if input.symbol == 'cETH':
+            token = Token(address=COMPOUND_ASSETS['WETH'])
+        elif input.address == '0xf5dce57282a584d2746faf1593d3121fcac444dc' and input.symbol == 'cDAI':
+            # When input = cSAI, it has been renamed to cDAI in the contract.
+            # We will still call up SAI
+            token = Token(address=COMPOUND_ASSETS['SAI'])
+        else:
+            token = Token(address=COMPOUND_ASSETS[input.symbol[1:]])
+
+        self.logger.info(f'{cToken.address, cToken.symbol}')
+
+        assert cToken.functions.isCToken().call()
+        # assert cToken.proxy_for is not None and cToken.functions.implementation().call() == cToken.proxy_for.address
+        assert cToken.functions.admin().call() == Address(COMPOUND_TIMELOCK)

```

Comment:
awesome, what's the best way to push this back to the data layer in the future? Hard to do without clear interpretation of the data from the data team.

Reply:
Let me start thinking about it. I learnt that data verification for accounts in inputs is critical.